### ISSUE TITLE
[ROADMAP][Updated on April 07] Megatron Core MoE Roadmap

### ISSUE THREAD

**Issue by yanring** [NVIDIA/Megatron-LM#1729]
## Description

The focus for Megatron Core MoE is to provide comprehensive support for latest MoE architectures, advanced parallelism strategies, and performance optimizations for Blackwell. This is a tentative roadmap and subject to change.

**🎉 This Roadmap is based on the [dev branch](https://github.com/NVIDIA/Megatron-LM/tree/dev); please see the details in its README.**

---

### Model Support

* ✅ **DeepSeek**  
  * ✅ DeepSeek-V2  
  * ✅ DeepSeek-V3, including MTP  
  * ✅ DeepSeek-V3.2 #2154 
* ✅ **Qwen**  
  * ✅ Qwen2-57B-A14B  
  * ✅ Qwen3-235B-A22B  
  * ✅ **(🚀New\!) Qwen3-Next** 
* ✅ **Mixtral**  

### Core MoE Functionality

* ✅ **Token dropless MoE** \- Advanced routing without token dropping  
* ✅ **Top-K Router** with flexible K selection  
* ✅ **Load balancing losses** for expert load balancing optimization

### Advanced Parallelism

* ✅ **Expert Parallel (EP)** with 3D parallelism integration  
* ✅ **Full parallelism combo**: EP \+ DP \+ TP \+ PP \+ SP support  
* ✅ **Context Parallel (CP)** for long sequence MoE training  
* ✅ **Parallel Folding** Heterogeneous Parallelism Mappings for Efficient Large-Scale MoE Model Training  
* ✅ **Distributed Optimizer for MoE** (ZeRO-1 equivalent)  
* ✅ **(🚀New\!) Megatron FSDP**/HSDP with full **expert parallel support**

### Optimizations

* ✅ **Memory Efficient token permutation**  
* ✅ **Fine-grained Recomputations** (mla, moe, mlp, moe\_act, norm)  
* ✅ **GroupedGEMM** and Gradient Accumulation Fusion  
* ✅ **DP/PP/TP/EP Communication Overlapping**  
* ✅ **Advanced fusions for Router; Permutation; MLA Rope; FP8 casting, etc**  
* ✅ **cuDNN fused Attention** and FlashAttn integration  
* ✅  (**🚀New\!) 1F1B EP A2A Overlap** \- Hiding Expert Parallel Communication with 1F1B Pipeline Schedule  
* ✅ **(🚀New\!) Muon and Layer-wise distributed optimizer** 
* ✅ **(🚀New\!) Pipeline-aware fine-grained activation offloading** #1912 
* ✅ **(🚀New\!) Production-ready cudaGraph support for MoE**

### Precision Support

* ✅ **GroupedGEMM** including FP8/MXFP8 support  
* ✅ **FP8 weights with BF16 optimizer states**  
* ✅ **FP8 training** full support

### Optimized Expert Parallel Communication Support

* ✅ **DeepEP support for H100 and B200**  
* ✅ **(🚀New\!) HybridEP for GB200** 

### Developer Experience

* ✅ [**MoE Model Zoo**](https://github.com/yanring/Megatron-MoE-ModelZoo) with pre-training best practices 
* ✅ **MCore2HF Converter** for ecosystem compatibility in megatron-bridge  
* ✅ **Distributed Checkpointing Support**  
* ✅ **Runtime Upcycling Support** for efficient model scaling  
* ✅ **Layer-wise logging** for detailed monitoring
---
### Performance & Kernel Optimizations

- [ ] **Split-K Indexer Kernels** - Avoid materializing [seqlen_q, seqlen_k] tensor with split-K kernels #2869 *(draft)*
- [x] **Absorbed MLA** - MLA computation optimization for DSA #3193
- [ ] HybridEP preprocess optimization
- [x] HybridEP fusion with permutation https://github.com/deepseek-ai/DeepEP/pull/588 https://github.com/NVIDIA/Megatron-LM/pull/4073/
- [x] mHC kernel fusion #3828
- [x] GEMM + SwiGLU fused MLP #3890 

### Long Context & Context Parallel

- [x] **Hybrid CP Part 2** - Enhanced hybrid data x context parallelism #2000
- [x] **THD Format E2E Support** - End-to-end THD format support #2924

### Model & Architecture

- [x] **Manifold Hyper Connection (mHC)** - #2943 
- [x] **GDN THD Support** - Packed sequence support for gated delta net #2644
- [x] **GDN Refinement** - Refine gated delta net implementation #3040

### Advanced Functionality

- [x] Router replay support for RL training (in progress) #2693
- [ ] Megatron FSDP performance optimization for MoE training

CUDA Graph Enhancements

- [ ] **MoE ECHO** - Elastic Cloning for sync-free, full CUDA-graph dropless MoE training #2368 *(draft)*
- [x] **Paged Stashing** - Dynamic tensor support for dropless MoE with CUDA graph #2690 *(draft)*
- [x] **CUDA Graph + Offloading** - Support CUDA Graph capture with offloading modules #3219 
- [ ] **Optimizer CUDA Graph** - Enable CUDA graph for ADAM optimizer #2931 *(draft)*
- [ ] CUDA Graph support for THD format(packed sequence) training.

## Ongoing Long-term Features

* **E2E Performance optimization** for DeepSeek-V3, Qwen-3 and other fine-grained MoEs
* **Sync-Free and Full-Iter cudaGraph MoE Training**
  * **MoE ECHO** for dropless MoE load balancing #2368
  * **Paged Stashing** for dynamic tensor support #2690
* **CPU Overhead Optimizations** for Blackwell Performance
* **MLA Optimizations**
  * **Absorbed MLA** #3044
  * **MLA CP 2.0** - MLA CP Enhancement for Longer Sequence Training
* THD and Long Context
	* **THD Format E2E Support** - End-to-end THD format support #2924
	* **Dynamic Context Parallel** for Imbalanced Long-Sequence Training
	  * #2054
	  * #2000
* **Megatron FSDP Performance Optimization for MoE Training**
* **Kernel fusions and optimizations for MoE models from TE** https://github.com/NVIDIA/TransformerEngine/issues/2438
* **New Architecture Support**
  * **Manifold Hyper Connection (mHC)** #2943
---

### Performance & Memory

- [x] **🚀 Fused Linear and Cross Entropy** - Fuse lm_head and CE to avoid materializing intermediate logits, reducing memory #2256
- [x] **🚀 Optimizer State Offloading** - Offload optimizer states and master weights to CPU for significant GPU memory savings #2987
- [x] **🚀 MTP Standalone Stages** - Support placing MTP layers into standalone pipeline stages for better VPP balance #1916
- [x] **🚀 DeepSeek V3.2 Support** - Performance optimizations for DeepSeek V3.2 #2154
- [x] DeepSeek V3 Pre-training Performance Guide (~960 TFLOPS/GPU on 256 GB200s) #1996

### CUDA Graph

- [x] **🚀 cuda_graph_scope Refactoring** #2353 #2694
- [x] **🚀 Partial CUDA Graph for EP Overlap** - Release CPU pressure within selected scope for EP A2A overlap #2810
- [x] TE cudagraph input memory optimization - Reuse static input memory buffer among microbatches #2391
- [x] CudaGraph compatibility with FP8 params (tensorwise & blockwise) #2087
- [x] NVFP4 MOE CUDA Graph support with 128 zero padding #2654

### Model & Parallelism

- [x] **🚀 Qwen3-Next Enhancements** - QK layernorm weight decay support and Gated Delta Net CP for long context #2825 #2614
- [x] **🚀 Hybrid Data x Context Parallelism** - New parallelism strategy combining DP and CP #2054
- [x] **🚀 Router Replay** - Deterministic routing mechanism for debugging and RL training #2693
- [x] **🚀 Fake Distributed Process Group** - Skip all distributed comm ops with `--fake-process-group` for profiling #2254
- [x] Remove padding token in MoE routing loss - Improve aux loss correctness and efficiency #2121
- [x] Context parallel support for eager attention implementation #1859
- [x] Packed sequence support in MTP module #2043

### Fine-grained Activation Offloading Enhancement

- [x] **🚀 OOP Refactoring** - Object-oriented redesign for fine-grained activation offloading #2905
- [x] Fix accuracy mismatch when offloading and recomputing same module #2123
- [x] Bug fix for fine-grained activation offloading in evaluate() #3041

### Megatron-FSDP

- [x] **🚀 FP8 Params Support** - MXFP8/Blockwise FP8 params for Megatron-FSDP #2828
- [x] **🚀 HSDP Support** - Hybrid Sharded Data Parallel with EP submesh registration #2388
- [x] Megatron-FSDP user guide documentation #2397

### Communication

- [x] **🚀 Hybrid-EP Upgrade** - Latest Hybrid-EP with kernel optimizations for EP64 and NVL8+IB #2424
- [x] HybridEP memory overhead reduction for 1F1B A2A overlap #2201

### Optimizer

- [x] **🚀 LayerWise DistOpt** - LayerWiseDistributedOptimizer with torch_dist checkpoint format and Muon support #1928 #2261

### Critical Bug Fixes

- [x] **Megatron-FSDP hang fix** - Resolve hang caused by non-deterministic reduce-scatter #2252
- [x] **EP Overlap correctness** - Fix missing final layernorm in EP overlap #2691
- [x] **Hybrid-EP hotfix** - Fix bug of hybrid-ep backend in flex-dispatcher #2287
- [x] **CUDA RNG Tracker** - Fix RNG tracker to use expert-parallel-rng correctly #2640
---

## Call for Community Contributions

* **Model implementations** \- Additional MoE model variants  
* **Performance testing** \- Performance tests across different platforms and workloads  
* **Documentation and tutorials** \- Best practices and optimization guides
* Bug fixes

---
This roadmap reflects the collective efforts of NVIDIA and our collaborators

Credits: MCore MoE Team and @sbhavani

**Labels:** `roadmap`, `moe`, `call-for-contribution`

**Comment by Skylion007** [comment]
No upstreaming of distributed MuON on the roadmap for Q3/Q4? Open to external contribution/PR?

**Comment by Skylion007** [comment]
Also FSDP2 has EP support now in PyTorch nightly and is used for such in TorchTitan. We probably should integrate that here too.

> Custom FSDP - NVIDIA-optimized FSDP implementation with full expert parallel support

**Comment by ericharper** [comment]
> No upstreaming of distributed MuON on the roadmap for Q3/Q4? Open to external contribution/PR?

This roadmap was for MoE only, we're planning on adding a more general roadmap for all of Core soon.

**Comment by jeromeku** [comment]
@yanring 

Very interested in learning more about the following:
- MCore2HF Converter
- Sync-Free MoE

Where can I find more info on these?

Thanks!

**Comment by switiz** [comment]
> > No upstreaming of distributed MuON on the roadmap for Q3/Q4? Open to external contribution/PR?
> 
> This roadmap was for MoE only, we're planning on adding a more general roadmap for all of Core soon.

Please consider distributed Muon support as a priority.
We are currently evaluating its performance across several open-source vendors such as Kimi and Zipu.

**Comment by sbhavani** [comment]
@switiz We are prioritizing Muon support and will share some of our early results soon

**Comment by yanring** [comment]
Hi @jeromeku, Thanks for reaching out. Some updates on your questions:

- MCore2HF Converter: Please check https://github.com/NVIDIA-NeMo/Megatron-Bridge
- Sync-Free MoE: A significant challenge with MoE is that dynamic shapes lead to D2H transfers and synchronization, which in turn creates considerable CPU overhead. We are currently developing a lossless solution to address this. While we can't disclose the specifics just yet, we hope to share more updates with you soon.

**Comment by yucc-leon** [comment]
Hi, what is the "mixture of linear attention"? MLA is famous for a variant of multi-head attention, and I can't find this "MLA" elsewhere using search engine...

**Comment by sbhavani** [comment]
@yucc-leon that's a typo, should be multi-head latent attention

**Comment by JazzeYoung** [comment]
> Hi, what is the "mixture of linear attention"? MLA is famous for a variant of multi-head attention, and I can't find this "MLA" elsewhere using search engine...

Is this Multi-Latent Attention?

**Comment by ChenXiaoTemp** [comment]
Will be qwen3-next-80B-A3b support?

**Comment by yanring** [comment]
Hey @ChenXiaoTemp, we've just merged qwen3-next support in the dev branch (https://github.com/NVIDIA/Megatron-LM/commit/c7590d8c3733619efa87a1a0733ac4cceedc683a) and modelzoo (https://github.com/yanring/Megatron-MoE-ModelZoo/pull/13). Please let us know if you run into any issues.

**Comment by andredido** [comment]
> Hi [@jeromeku](https://github.com/jeromeku), Thanks for reaching out. Some updates on your questions:
> 
> * MCore2HF Converter: Please check https://github.com/NVIDIA-NeMo/Megatron-Bridge
> * Sync-Free MoE: A significant challenge with MoE is that dynamic shapes lead to D2H transfers and synchronization, which in turn creates considerable CPU overhead. We are currently developing a lossless solution to address this. While we can't disclose the specifics just yet, we hope to share more updates with you soon.

Hi @yanring, I am trying to convert a checkpoint trained with Megatron‑LM (of a Qwen3-30B-A3B model). I wasn’t able to complete the conversion. Are there any official documents or guides for converting Megatron-LM models via the Megatron‑Bridge tool?

**Comment by sbhavani** [comment]
@andredido please file an issue in Megatron Bridge repo and we'll take a look. You can find some documentation [here](https://github.com/NVIDIA-NeMo/Megatron-Bridge/tree/main/src/megatron/bridge/models#converting-megatron-models-back-to-hugging-face).

**Comment by SeanLi-OI** [comment]
> Hi [@jeromeku](https://github.com/jeromeku), Thanks for reaching out. Some updates on your questions:
> 
> * MCore2HF Converter: Please check https://github.com/NVIDIA-NeMo/Megatron-Bridge
> * Sync-Free MoE: A significant challenge with MoE is that dynamic shapes lead to D2H transfers and synchronization, which in turn creates considerable CPU overhead. We are currently developing a lossless solution to address this. While we can't disclose the specifics just yet, we hope to share more updates with you soon.

Hi @yanring , I'm wondering if sync-free MoE has compatibility with dropless MoE? Or it's only support MoE with token drop?

**Comment by yanring** [comment]
> Hi [@yanring](https://github.com/yanring) , I'm wondering if sync-free MoE has compatibility with dropless MoE? Or it's only support MoE with token drop?

Yeah, we are targeting dropless MoE.

**Comment by yanring** [comment]
📢 Hey everyone! We just updated the roadmap, our current target is v0.16. Feel free to drop any questions or thoughts!

**Comment by yeontaek** [comment]
Hi @yanring 

It’s currently released up to version 0.14.0, so is it correct to assume that the target is to reach version 0.16.0 by Q4?

Also, could you share what features or improvements are planned for version 0.15.0?

**Comment by sbhavani** [comment]
You can follow along Transformer Engine optimizations for MoE here: https://github.com/NVIDIA/TransformerEngine/issues/2438

**Comment by x41lakazam** [comment]
Hi,

What is the meaning of "DP/PP/TP/EP Communication Overlapping"?

**Comment by yanring** [comment]
> Hi,
> 
> What is the meaning of "DP/PP/TP/EP Communication Overlapping"?

@x41lakazam  In Megatron Core, we've overlapped communication and computation for all four of these parallelism types:

| Type | Parallelism | Communication Being Overlapped |
|------|-------------|-------------------------------|
| **DP** | Data Parallel | Gradient All-Reduce/Reduce-Scatter, Parameter All-Gather |
| **PP** | Pipeline Parallel | P2P Send/Recv between pipeline stages |
| **TP** | Tensor Parallel | Activation All-Gather/Reduce-Scatter |
| **EP** | Expert Parallel | All-to-All token dispatch communication |

#### DP Communication Overlap
```bash
--overlap-grad-reduce        # Overlap gradient reduce with backward computation
--overlap-param-gather       # Overlap param all-gather with forward (distributed optimizer)
--overlap-param-gather-with-optimizer-step  # Overlap with optimizer step
```

#### PP Communication Overlap
```bash
--overlap-p2p-comm           # Enabled by default with VPP; overlaps P2P with forward/backward chunks
--overlap-p2p-communication-warmup-flush  # Also overlap during warmup and flush phases
```

#### TP Communication Overlap
```bash
--tp-comm-overlap            # Enable TP comm overlap with GEMM kernels
--tp-comm-overlap-cfg        # Config file for overlap settings
--disable-tp-comm-overlap-ag # Disable All-Gather + GEMM pipelining
--disable-tp-comm-overlap-rs # Disable Reduce-Scatter + GEMM pipelining
--tp-comm-overlap-rs-dgrad   # Enable RS overlap with dgrad GEMM
--disable-tp-comm-bulk-dgrad # Disable AG overlap with activation gradient GEMM
--disable-tp-comm-bulk-wgrad # Disable RS overlap with weight gradient GEMM
```

#### EP Communication Overlap (MoE)
```bash
--overlap-moe-expert-parallel-comm  # Batch-level overlap to hide EP A2A communication
--delay-wgrad-compute               # Delay wgrad compute for batch-level overlapping
--ep-overlap-early-attn-memory-release  # Release attention memory early in EP overlap
```

**Core idea**: Use async communication + compute/comm pipelining so the GPU continues computation while waiting for communication to complete, maximizing hardware utilization.

**Comment by chenhongyu2048** [comment]
Hello, I noticed that FSDP+EP support was mentioned in https://github.com/NVIDIA/Megatron-LM/pull/2388. I would like to ask if this means that the weights of an expert will be further split across multiple `dp_replicates`, in a manner similar to Zero3? @shjwudp 

Additionally, I noticed that the current roadmap mentions a need for
"Megatron FSDP Performance Optimization for MoE Training." I am very interested in this. Could you please tell me how I can learn more about it and whether it is possible for me to participate in this part of the work? @yanring

**Comment by sbhavani** [comment]
This roadmap has been superseded by the current Megatron Core MoE roadmap: #4815. Unfinished work remains tracked through that roadmap and its individual issues or pull requests.
